bot/exts/backend/branding/_repository.py

```diff
@@ -0,0 +1,240 @@
+    We work with the assumption that the branding repository checks for such conflicts and prevents them
+    from reaching the main branch.
+
+    This class keeps no internal state. All `get_current_event` calls will result in GitHub API requests.
+    The caller is therefore responsible for being responsible and caching information to prevent API abuse.
+
+    Requests are made using the HTTP session looked up on the bot instance.
+    """
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+
+    async def fetch_directory(self, path: str, types: t.Container[str] = ("file", "dir")) -> t.Dict[str, RemoteObject]:
+        """
+        Fetch directory found at `path` in the branding repository.
+
+        The directory will be represented by a mapping from file or sub-directory names to their corresponding
+        instances of `RemoteObject`. Passing a custom `types` value allows only getting files or directories.
+
+        An exception will be raised if the request fails, or if the response lacks the expected keys.
+        """
+        full_url = f"{BRANDING_URL}/{path}"
+        log.debug(f"Fetching directory from branding repository: {full_url}")
+
+        async with self.bot.http_session.get(full_url, params=PARAMS, headers=HEADERS) as response:
+            if response.status != 200:
+                raise RuntimeError(f"Failed to fetch directory due to status: {response.status}")
+            json_directory = await response.json()
+
+        return {file["name"]: RemoteObject(file) for file in json_directory if file["type"] in types}
+
+    async def fetch_file(self, download_url: str) -> bytes:
+        """
+        Fetch file as bytes from `download_url`.
+
+        Raise an exception if the request does not succeed.
+        """
+        log.debug(f"Fetching file from branding repository: {download_url}")
+
+        async with self.bot.http_session.get(download_url, params=PARAMS, headers=HEADERS) as response:
+            if response.status != 200:
+                raise RuntimeError(f"Failed to fetch file due to status: {response.status}")
+            return await response.read()
+
+    def parse_meta_file(self, raw_file: bytes) -> MetaFile:
+        """
+        Parse a 'meta.md' file from raw bytes.
+
+        The caller is responsible for handling errors caused by misconfiguration.
+        """
+        attrs, description = frontmatter.parse(raw_file, encoding="UTF-8")
+
+        if not description:
+            raise BrandingMisconfiguration("No description found in 'meta.md'!")
+
+        if attrs.get("fallback", False):
+            return MetaFile(is_fallback=True, start_date=None, end_date=None, description=description)
+
+        start_date_raw = attrs.get("start_date")
+        end_date_raw = attrs.get("end_date")
+
+        if None in (start_date_raw, end_date_raw):
+            raise BrandingMisconfiguration("Non-fallback event doesn't have start and end dates defined!")
```

Comment:
Even if it's validated branding-side, bot still needs to do something in the case that the dates aren't there. In either case, we'll get an exception; in this case, the exception will actually tell you the cause of the error.

Despite branding having validation, you can still get an invalid event here e.g. if someone pushes directly to branding main. CI will fail, but we fetch directly from the repo. Or what if there's a bug in the validation script? It's unlikely, but an exception must be raised either way, and since I can explain to whoever will read it what happened, I might as well.

The same justification applies for checking for missing assets, and informing the maintainers exactly what's missing in which event, rather than hoping they'll figure it out from the key / value error. The problem with those is that they look like bugs.

How do you feel about this?

Reply:
I agree with the reasons, so I think this should be good!